### PR TITLE
expression: make some datetime functions' behavior consistent with MySQL

### LINKED ISSUE #30683: inconsistent result of `quarter` when the input is zero date
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
mysql> desc datetime_test;
+-------+-------------+------+------+---------+-------+
| Field | Type        | Null | Key  | Default | Extra |
+-------+-------------+------+------+---------+-------+
| id    | int(11)     | YES  |      | NULL    |       |
| v1    | datetime    | YES  |      | NULL    |       |
| v2    | datetime(3) | YES  |      | NULL    |       |
+-------+-------------+------+------+---------+-------+
3 rows in set (0.00 sec)

mysql> select v1,v2,quarter(v1), quarter(v2) from datetime_test;
+---------------------+-------------------------+-------------+-------------+
| v1                  | v2                      | quarter(v1) | quarter(v2) |
+---------------------+-------------------------+-------------+-------------+
| 0000-00-00 00:00:00 | 0000-00-00 00:00:00.000 |           0 |        NULL |
+---------------------+-------------------------+-------------+-------------+
1 row in set, 1 warning (0.01 sec)
```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
The result of `quarter(v1)` and `quarter(v2)` should be the same, both 0 or both null.
BTW in MySQL, both `quarter(v1)` and `quarter(v2)` return 0.
### 3. What did you see instead (Required)
`quarter(v1)` returns 0, and `quarter(v2)` returns null
### 4. What is your TiDB version? (Required)
mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                            |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.4.0-alpha-334-g376d5428c
Edition: Community
Git Commit Hash: 376d5428cca95ae51143162ee0f7f3972e84cff7
Git Branch: master
UTC Build Time: 2021-12-07 07:30:04
GoVersion: go1.17.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by lzmhhh123**
Signed-off-by: lzmhhh123 <lzmhhh123@gmail.com>

### What problem does this PR solve?

Issue Number: close #30683

Problem Summary:

### Check List

Tests 

- [X] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [X] Changes MySQL compatibility

### Release note

```release-note
Make some DateTime functions' behavior consistent with MySQL.
```

### CODE REVIEW

File: expression/integration_serial_test.go [hunk 1]
```diff
@@ -2043,10 +2043,8 @@ func TestTimeBuiltin(t *testing.T) {
 	result = tk.MustQuery(`select quarter("2012-14-20"), quarter("aa"), quarter(null), quarter(11), quarter(12.99);`)
 	result.Check(testkit.Rows("<nil> <nil> <nil> <nil> <nil>"))
 	result = tk.MustQuery(`select quarter("0000-00-00"), quarter("0000-00-00 00:00:00");`)
-	result.Check(testkit.Rows("<nil> <nil>"))
-	tk.MustQuery("show warnings").Check(testutil.RowsWithSep("|",
-		"Warning|1292|Incorrect datetime value: '0000-00-00 00:00:00.000000'",
-		"Warning|1292|Incorrect datetime value: '0000-00-00 00:00:00.000000'"))
+	result.Check(testkit.Rows("0 0"))
```
Comment (windtalker) [inline expression/integration_serial_test.go:2046]: This test is running under allow zero date sql mode?

Comment (lzmhhh123) [reply expression/integration_serial_test.go:2046]: Yes, see here https://github.com/pingcap/tidb/pull/31078/files#diff-3b513c8044c3b047af40bb879b7e650a636ee695d5740881f8d861c52fa96c60L1990

**Review by windtalker — APPROVED** [review]
LGTM

**Review by ichn-hu — APPROVED** [review]

**Pull request merged**